### ISSUE TITLE
Refactor TribitsCtestDriverCore.cmake to allow running steps in different executions

### ISSUE THREAD

**Issue by bartlettroscoe** [TriBITSPub/TriBITS#259]
**CC:** @fryeguy52 

## Description

Ctest -S driver scripts written using `TribitsCtestDriverCore.cmake` currently always call `CTEST_START()`, `CTEST_CONFGIURE()` and `CTEST_BUILD()`, everytime they run.  By default, it will also call `CTEST_TEST()` and other test-related commands.

This setup has been find up till now, but some of the Trilinos ATDM builds require doing the configure and build on one node and to have the tests run on another node.  This is the case for the SNL HPC nodes.  We need to find some way to address this.

One way to address this would be to have every test that ctest ran actaully be a seprate job submission script.  But that would create a huge number of jobs in the queue and could really take a long time to get all of these through.  And it would take refactoring the TriBITS Core software to wrap every TRIBTS_ADD_TEST() and TRIBITS_ADD_ADVANCED_TEST() invocation that was run by `ADD_TEST()`.  That is essentially what the SPARC custom CMake/CTest system can do.

Another way to accomidate this would be to try to allow the `ctest -S <script>.cmake` command to be called twice; the first time to do `CTEST_START()` + `CTEST_UPDATE()` + `CTEST_CONFIGURE()` + `CTEST_BUILD()` and then a second time to run `CTEST_TEST()` and perhaps `CTEST_COVERAGE()`, `CTEST_MEMCHECK()`, once to run the tests.  But this would only really work for the all-at-once mode, not the package-by-package mode.  Because the package-by-package mode runs the tests for each package individuall before moving on to the next.  If this were implemented, then one could run an all-at-once update, configure, build, test, and summit as:

```
$ env  CTEST_DO_TEST=OFF ctest -S <script>.cmake

$ env  CTEST_DO_START=OFF  CTEST_DO_CONFIGURE=OFF  CTEST_DO_BUILD=OFF  ctest -S <script>.cmake
```

**Comment by bartlettroscoe** [comment]
The branch:

* https://github.com/bartlettroscoe/TriBITS/tree/259-two-step-ctest-s-scripts

contains the basics of doing seprated ctest -S commands submitting to the same CDash build.  Turns out that `ctest_start(APPEND)` can be called to append results for an existing build.

I did some testing below for the all-at-once mode including submitting to CDash for the TribitExProj and Trilinos and it looks to be doing the right thing.

Next, I will add some more test cases and try to get this working for the package-by-package mode as well (why not).

<details>
<summary>
  <b>DETAILED NOTES:</b> (click to expand)
</summary>

I pushed the updated branch:

```
To git@github.com:bartlettroscoe/TriBITS.git
 + ef8fbc3...b5e2571 259-two-step-ctest-s-scripts -> 259-two-step-ctest-s-scripts (forced update)
```

With that, a two-stage ctest -S invocations with the all-at-once mode submit to the same CDAsh project.  I have a good unit test for this first two-stage ctest -S case for the all-at-once appraoch.  Using CMake 3.11.1 that I installed for this, using the TriBITS do-configure script on crf450:

```
#!/bin/bash

../../../Trilinos/TriBITS/dev_testing/generic/do-configure-mpi-debug \
-DCTEST_PARALLEL_LEVEL=16 \
-DTriBITS_ENABLE_CONFIGURE_TIMING=ON \
-DDART_TESTING_TIMEOUT=60 \
"$@"

#-DTriBITS_CTEST_DRIVER_COVERAGE_TESTS=TRUE \
#-DTriBITS_CTEST_DRIVER_MEMORY_TESTS=TRUE \
#-DTribitsExProj_CTEST_USE_NEW_AAO_FEATURES=TRUE \

# NOTE: Add -DTriBITS_CTEST_DRIVER_SUBMIT_TO=EXP_CDASH to submit to
# exp.cdash.org

# NOTE: Add -DTriBITS_CTEST_DRIVER_SUBMIT_TO=TESTING_VM_SANDIA_CDASH to submit
# to testing-vm.sandia.gov
```

I did the configure and submit of this unit test on my machine crf450 with:

```
$ cd ~/Trilinos.base/BUILDS/GCC-4.8.3/TRIBITS_MPI_DEBUG/

$  export PATH=/home/vera_env/common_tools/cmake-3.11.1/bin:$PATH

$ ./do-configure -DTribitsExProj_CTEST_USE_NEW_AAO_FEATURES=TRUE \
  -DTriBITS_CTEST_DRIVER_SUBMIT_TO=TESTING_VM_SANDIA_CDASH

$ make

$ ctest -VV -R TriBITS_CTestDriver_AAO_ST_ALL_PASS_CALLS_2 &> console.out
```

That submitted to:

* https://testing-vm.sandia.gov/cdash/index.php?project=TribitsExProj&date=2018-04-26&filtercombine=and&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=TriBITS_CTestDriver_AAO_ST_ALL_PASS_CALLS_2_SAME_BUILD&field2=buildstamp&compare2=61&value2=20180426-1323-Continuous

That shows that it all showed up in one build.

But that even worked when I did two seprate calls to `ctest_start(...)` that started new CDash "dashboards" since the build samps were the same because they started in the same minute!

To verified this with an all-at-once configure, build, test, and submit works with a longer build, I used the 'dashboard' target with Trilinos for this branch of TriBITS.

Using the do-configure script:

```
#!/bin/bash
cmake \
-DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/MpiReleaseDebugSharedPtSerial.cmake \
-DTrilinos_ENABLE_CONFIGURE_TIMING=ON \
-DTrilinos_ENABLE_TESTS=ON \
-DCTEST_BUILD_FLAGS=-j16 \
-DCTEST_PARALLEL_LEVEL=16 \
"$@" \
../../../Trilinos
```

I did the following with Trilinos that I did with:

```
$ cd ~/Trilinos.base/BUILDS/GCC-4.8.4/MPI_RELEASE_DEBUG_SHARED_PT/

$ rm -r CMake*

$ . ~/Trilinos.base/Trilinos/cmake/load_sems_dev_env.sh

$ export PATH=/home/vera_env/common_tools/cmake-3.11.1/bin:$PATH

$ ./do-configure -DTrilinos_TRIBITS_DIR:STRING=TriBITS/tribits \
  -DTrilinos_CTEST_DO_ALL_AT_ONCE=ON -DTrilinos_CTEST_USE_NEW_AAO_FEATURES=TRUE \
  -DTrilinos_ENABLE_Teuchos=ON \
  &> configure.out

$ env CTEST_DO_TEST=OFF \
  make dashboard &> make.dashboard.1.out

$ env CTEST_DO_NEW_START=OFF CTEST_DO_CONFIGURE=OFF CTEST_DO_BUILD=OFF \
  make dashboard &> make.dashboard.2.out
```

That submitted to:

* https://testing-vm.sandia.gov/cdash/index.php?project=Trilinos&filtercombine=and&filtercombine=and&filtercombine=and&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Linux-MPI_RELEASE_DEBUG_SHARED_PT&field2=buildstamp&compare2=61&value2=20180426-1507-Experimental&field3=site&compare3=61&value3=crf450.srn.sandia.gov

That looks good.

</details>

**Comment by bartlettroscoe** [comment]
The updated branch:

```
To git@github.com:bartlettroscoe/TriBITS.git
 + 65d64d9...4a1c459 259-two-step-ctest-s-scripts -> 259-two-step-ctest-s-scripts (forced update)
```

totally has this working for both the all-at-once and package-by-package approaches.  I was surprised this worked for the package-by-package approach.

Now I just need to add documentation for this new feature and then I can merge this to 'master'.

**Comment by bartlettroscoe** [comment]
This is complete with the merge of PR #260 and the fix in c1665e0.

Now to use this in the ATDM Trilinos builds where it is needed (see [TRIL-204](https://software-sandbox.sandia.gov/jira/browse/TRIL-204)).
